xcode/Subconscious/Shared/Models/Slug.swift

```diff
@@ -2,92 +2,79 @@
 //
 
 import Foundation
 
-/// A slug is a normalized identifier (basically "words-and-dashes")
+/// A type representing a valid slug (`/slug`)
 struct Slug:
-    Identifiable,
     Hashable,
     Equatable,
+    Identifiable,
     Comparable,
-    LosslessStringConvertible,
-    Codable
+    Codable,
+    LosslessStringConvertible
 {
-    /// Compare slugs by alpha
+    private static let slugRegex = /([\w\d\-]+)(\/[\w\d\-]+)*/
+
     static func < (lhs: Slug, rhs: Slug) -> Bool {
         lhs.id < rhs.id
     }
     
-    /// Sanitize a string into a "slug string"-a string into a string that can
-    /// be losslessly converted to and from a Slug.
-    ///
-    /// If a valid slug string cannot be produced, returns nil.
-    ///
-    /// If you need a value to truly be a slug, use the Slug constructor.
-    static func format(_ string: String) -> String? {
+    static func == (lhs: Self, rhs: Self) -> Bool {
+        lhs.id == rhs.id
+    }
+    
+    /// Attempt to sanitize a string into a "slug string"-a string that
+    /// represents a valid slug.
+    private static func format(_ string: String) -> String {
         // Strip all non-allowed characters
-        let slugString = string.replacingOccurrences(
-            of: #"[^a-zA-Z0-9_\-\/\s]"#,
+        let formatted = string.replacingOccurrences(
+            of: #"[^\w\d\-\s]"#,
             with: "",
             options: .regularExpression,
             range: nil
         )
         // Trim leading/trailing whitespace
         .trimmingCharacters(in: .whitespacesAndNewlines)
-        // Then trim leading/trailing slashes
-        .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
-        .lowercased()
         // Replace runs of one or more space with a single dash
         .replacingOccurrences(
             of: #"\s+"#,
             with: "-",
             options: .regularExpression,
             range: nil
         )
-        // Replace all instances of two or more consecutive slashes
-        // with a single slash.
-        .replacingOccurrences(
-            of: #"/+"#,
-            with: "/",
-            options: .regularExpression,
-            range: nil
-        )
         .truncatingSafeFileNameLength()
-        guard !slugString.isEmpty else {
-            return nil
-        }
-        return slugString
+        return formatted
     }
     
-    let id: String
-    var description: String { id }
+    let description: String
+    let verbatim: String
+
+    var id: String { description }
+
+    var markup: String {
+        "/\(verbatim)"
+    }
     
     /// Losslessly create a slug from a string.
     /// This requires that the string already be formatted like a
-    /// sanitized slug, including being lowercased.
-    init?(_ string: String) {
-        guard let id = Self.format(string) else {
+    /// valid slug.
+    init?(_ description: String) {
+        guard description.wholeMatch(of: Self.slugRegex) != nil else {
```

Comment:
Stricter and safer parsing via regular expression.

Previously we were formatting the string and checking if the formatted version matched the input. However, formatting code was just a series of transformations. A regular language lets us express our invariants in a way that is correct and complete.